src/data/storage.ts

```diff
@@ -0,0 +1,130 @@
+import * as idb from "idb-keyval";
+import { ExcalidrawElement } from "../element/types";
+import { AppState } from "../types";
+import { clearAppStateForLocalStorage } from "../appState";
+import { restore } from "./restore";
+
+const STORAGE_NAME = "excalidraw";
+const STORAGE_KEY_ELEMENTS = "excalidraw";
+const STORAGE_KEY_STATE = "excalidraw-state";
+const STORAGE_KEY_COLLAB = "excalidraw-collab";
+
+/**
+ * Uses indexedDB as key-value storage.
+ * Fallsback to localStorage on older browsers without IDB (or) while running tests.
+ */
+export class WebStorageProvider {
+  private supportsIDB: boolean;
+  private indexedDBStore: idb.Store | null;
+
+  constructor() {
+    this.supportsIDB = "indexedDB" in window && process.env.NODE_ENV !== "test";
+    this.indexedDBStore = this.supportsIDB
+      ? new idb.Store(`${STORAGE_NAME}-db`, `${STORAGE_NAME}-store`)
+      : null;
+  }
+
+  async clear(): Promise<void> {
+    if (this.supportsIDB && this.indexedDBStore) {
+      return await idb.clear(this.indexedDBStore);
+    }
+    return localStorage.clear();
+  }
+
+  async delete(key: string): Promise<void> {
+    if (this.supportsIDB && this.indexedDBStore) {
+      return await idb.del(key, this.indexedDBStore);
+    }
+    return localStorage.removeItem(key);
+  }
+
+  async get(key: string): Promise<string | null> {
+    if (this.supportsIDB && this.indexedDBStore) {
+      return await idb.get(key, this.indexedDBStore);
+    }
+    return localStorage.getItem(key);
+  }
+
+  async set(key: string, value: string): Promise<void> {
+    if (this.supportsIDB && this.indexedDBStore) {
+      return await idb.set(key, value, this.indexedDBStore);
+    }
+    return localStorage.setItem(key, value);
+  }
+}
+
+const storage = new WebStorageProvider();
+
+export const saveUsernameToStorage = async (username: string) => {
+  try {
+    await storage.set(STORAGE_KEY_COLLAB, JSON.stringify({ username }));
+  } catch (error) {
+    console.error(error);
+  }
+};
+
+export const restoreUsernameFromStorage = async (): Promise<string | null> => {
+  try {
+    const data = await storage.get(STORAGE_KEY_COLLAB);
+    if (data) {
+      return JSON.parse(data).username;
+    }
+  } catch (error) {
+    console.error(error);
+  }
+
+  return null;
+};
+
+export const saveToStorage = async (
+  elements: readonly ExcalidrawElement[],
+  appState: AppState,
+) => {
+  try {
+    await storage.set(
+      STORAGE_KEY_ELEMENTS,
+      JSON.stringify(elements.filter((element) => !element.isDeleted)),
```

Comment:
As it's possible to store objects in indexedDB, can we use `JSON.stringify` only in `localStorage` implementation?

Reply:
IndexedDB clones first, but from what we've seen lately when working with history, cloning might be faster than serialization. Unless the [structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) is way slower than what we do.

Reply:
`JSON.parse(JSON.stringify())` is faster, compared to the structured clone algorithm, but not by much. In real world scenarios, this should not ideally matter. The advantage here of directly storing in objects, is the ability to support cyclic references, that JSON.stringify doesnt support.

@dwelle This [article](https://dassur.ma/things/deep-copy/) has performance benchmarks, comparing JSON.parse to implementations ( like postMessage, history API ) that use Structured Clone Implementation.

Reply:
@dwelle About this, should we pursue the idea of storing objects, or no?

Reply:
> `JSON.parse(JSON.stringify())` is faster, compared to the structured clone algorithm, but not by much. In real world scenarios, this should not ideally matter.

If you refer to the benchmark in the article, then that's dubious because they're using other APIs to get access to the structured clone algo.

> The main advantage here in directly storing as objects, is the ability to support cyclic references, that JSON.stringify doesnt support, but I don't that would be of much value to us.

Interesting, didn't know. Just tested via the `MessageChannel` hack. But does that work for IDB, too? Either way, we wouldn't be able to depend on this, anyway, because IDB isn't the only place where we clone state (we'd have to use structured clone or similar everywhere).

Reply:
> If you refer to the benchmark in the article, then that's dubious because they're using other APIs to get access to the structured clone algo.

Agreed, I couldn't really find other benchmarks, that are meant to performance test only, sttuctured clone as it's not exposed. I'll try to find out if something else exists, in the context of IDB and structured clone.

Reply:
But then again, we'd only be doing serialization (`JSON.stringify`), while they were deserializing in the same step.

Anyway, I wouldn't worry too much about it. We can swap the implem with relative ease (the migration would consist of checking if the IDB returned a string, or an object).

Reply:
> Yup it works with IDB also, you can see an example here.

They state:

> An example of object that can’t be stored: an object with circular references.

Reply:
@dwelle Oops, my bad. I read it wrongly: "The indexeddb is only capable of storing data that doesn't have circular references."